streams/src/test/java/org/apache/kafka/streams/kstream/JoinWindowsTest.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.kafka.streams.kstream;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+
+
+public class JoinWindowsTest {
+
+    private static String anyName = "window";
+    private static long anySize = 123L;
+    private static long anyOtherSize = 456L;
+
+    @Test
+    public void shouldHaveSaneEqualsAndHashCode() {
+        JoinWindows w1 = JoinWindows.of("w1", anySize);
+        JoinWindows w2 = JoinWindows.of("w2", anySize);
+
+        // Reflexive
+        assertEquals(w1, w1);
+        assertEquals(w1.hashCode(), w1.hashCode());
+
+        // Symmetric
+        assertEquals(w1, w2);
+        assertEquals(w2, w1);
+        assertEquals(w1.hashCode(), w2.hashCode());
+
+        JoinWindows w3 = JoinWindows.of("w3", w2.after).before(anyOtherSize);
+        JoinWindows w4 = JoinWindows.of("w4", anyOtherSize).after(w2.after);
+        assertEquals(w3, w4);
+        assertEquals(w4, w3);
+        assertEquals(w3.hashCode(), w4.hashCode());
+
+        // Inequality scenarios
+        assertNotEquals("must be false for null", null, w1);
```

Comment:
The way this is written is confusing.

I prefer the style in `TimeWindowsTest`:

```java
assertFalse("must be false for null", w1.equals(null));
```
Why?  Because we are validating the behavior of `JoinWindows#equals()` here, so we should actually call the method `equals()` directly, rather than having the unit testing framework call it indirectly via `assertNotEquals`.  Also, the assert message "must be false for null" makes sense for the `assertFalse` variant above, but it is confusing for the `assertNoEquals` variant.